### PR TITLE
support sign and auth for request

### PR DESCRIPTION
**Pull request by CodingSinger**
enhance request security, support custom signing and verification.
refer https://github.com/apache/dubbo/issues/5461

### CODE REVIEW

**Comment by CodingSinger** [comment]
Please don't merge, I will add unit tests later.

File: filter/filter_impl/auth/authenticator_test.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	"strconv"
+	"testing"
+	"time"
+)
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/protocol/invocation"
+	"github.com/stretchr/testify/assert"
```
Comment (flycash) [inline filter/filter_impl/auth/authenticator_test.go]: format the imports

File: filter/filter_impl/auth/sign_util.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	signature := mac.Sum(nil)
+	return base64.URLEncoding.EncodeToString(signature)
+}
+
+func IsEmpty(s string, allowSpace bool) bool {
+	if len(s) == 0 {
+		return true
+	}
+	if !allowSpace {
+		if strings.TrimSpace(s) == "" {
```
Comment (flycash) [inline filter/filter_impl/auth/sign_util.go]: return strings.TrimeSpace(s) == ""

File: filter/filter_impl/auth/accesskey_storage_test.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import (
+	"net/url"
+	"testing"
+)
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/constant"
+	invocation2 "github.com/apache/dubbo-go/protocol/invocation"
+	"github.com/stretchr/testify/assert"
```
Comment (zouyx) [inline filter/filter_impl/auth/accesskey_storage_test.go]: split package

File: filter/filter_impl/auth/consumer_sign_test.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/protocol"
+	"github.com/apache/dubbo-go/protocol/invocation"
+	"github.com/apache/dubbo-go/protocol/mock"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
```
Comment (zouyx) [inline filter/filter_impl/auth/consumer_sign_test.go]: split 2 packages

File: filter/filter_impl/auth/consumer_sign.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+func init() {
+	extension.SetFilter(constant.CONSUMER_SIGN_FILTER, getConsumerSignFilter)
+}
+
+func (filter *ConsumerSignFilter) Invoke(invoker protocol.Invoker, invocation protocol.Invocation) protocol.Result {
+	logger.Infof("invoking ConsumerSign filter.")
+	url := invoker.GetUrl()
+	shouldAuth := url.GetParamBool(constant.SERVICE_AUTH_KEY, false)
```
Comment (zouyx) [inline filter/filter_impl/auth/consumer_sign.go]: should u extract common method for auth? ProviderAuthFilter seems has the same logic

File: filter/filter_impl/auth/provider_auth.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+func init() {
+	extension.SetFilter(constant.PROVIDER_AUTH_FILTER, getProviderAuthFilter)
+}
+
+func (filter *ProviderAuthFilter) Invoke(invoker protocol.Invoker, invocation protocol.Invocation) protocol.Result {
+	logger.Infof("invoking providerAuth filter.")
+	url := invoker.GetUrl()
+	shouldAuth := url.GetParamBool(constant.SERVICE_AUTH_KEY, false)
+	if shouldAuth {
```
Comment (zouyx) [inline filter/filter_impl/auth/provider_auth.go]: it seems the same logic to ConsumerSignFilter

File: filter/filter_impl/auth/provider_auth_test.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	"time"
+)
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/protocol"
+	"github.com/apache/dubbo-go/protocol/invocation"
+	"github.com/apache/dubbo-go/protocol/mock"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
```
Comment (zouyx) [inline filter/filter_impl/auth/provider_auth_test.go]: split 2 packages

File: filter/filter_impl/auth/provider_auth.go (hunk 6 above)
Comment (CodingSinger) [reply filter/filter_impl/auth/provider_auth.go]: I don't think this is necessary. The consumer side calls the sign method, but the provider side calls the auth method, and the error handling is not the same. This is an overwrap and does not make the code more readable. Or do you have a better way?

File: filter/filter_impl/auth/sign_util.go (hunk 2 above)
Comment (CodingSinger) [reply filter/filter_impl/auth/sign_util.go]: Good catch, I will take it.

File: common/url.go [hunk 8]
```diff
@@ -313,6 +313,26 @@ func (c URL) ServiceKey() string {
 	return buf.String()
 }
 
+func (c *URL) ColonSeparatedKey() string {
```
[the remark this answers is not in the document]
Comment (CodingSinger) [reply common/url.go]: emmm... This follows the name of dubbo java...

File: filter/filter_impl/auth/authenticator_test.go (hunk 1 above)
Comment (CodingSinger) [reply filter/filter_impl/auth/authenticator_test.go]: done.

File: filter/filter_impl/auth/accesskey_storage_test.go (hunk 3 above)
Comment (CodingSinger) [reply filter/filter_impl/auth/accesskey_storage_test.go]: done.

File: filter/filter_impl/auth/consumer_sign_test.go (hunk 4 above)
Comment (CodingSinger) [reply filter/filter_impl/auth/consumer_sign_test.go]: done.

File: filter/filter_impl/auth/provider_auth_test.go (hunk 7 above)
Comment (CodingSinger) [reply filter/filter_impl/auth/provider_auth_test.go]: done.

File: filter/filter_impl/auth/authenticator.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+func (authenticator *DefaultAuthenticator) Authenticate(invocation protocol.Invocation, url *common.URL) error {
+	accessKeyId := invocation.AttachmentsByKey(constant.AK_KEY, "")
+
+	requestTimestamp := invocation.AttachmentsByKey(constant.REQUEST_TIMESTAMP_KEY, "")
+	originSignature := invocation.AttachmentsByKey(constant.REQUEST_SIGNATURE_KEY, "")
+	consumer := invocation.AttachmentsByKey(constant.CONSUMER, "")
+	if IsEmpty(accessKeyId, false) || IsEmpty(consumer, false) ||
+		IsEmpty(requestTimestamp, false) || IsEmpty(originSignature, false) {
+		return errors.New("failed to authenticate, maybe consumer not enable the auth")
```
[the remark this answers is not in the document]
Comment (CodingSinger) [reply filter/filter_impl/auth/authenticator.go]: done.

File: filter/filter_impl/auth/provider_auth_test.go (hunk 7 above)
Comment (flycash) [reply filter/filter_impl/auth/provider_auth_test.go]:
```go
import (
   SDK's library
)

import (
  third party's library
)

import (
   dubbo-go's library
)
```

File: filter/filter_impl/auth/consumer_sign.go (hunk 5 above)
Comment (flycash) [reply filter/filter_impl/auth/consumer_sign.go]: It's hard to do that, I think.

File: filter/filter_impl/auth/provider_auth_test.go (hunk 7 above)
Comment (CodingSinger) [reply filter/filter_impl/auth/provider_auth_test.go]: I have resolve it.  Why can't see updates here?

File: filter/filter_impl/auth/consumer_sign.go (hunk 5 above)
Comment (CodingSinger) [reply filter/filter_impl/auth/consumer_sign.go]: FIXED.

File: filter/filter_impl/auth/provider_auth_test.go (hunk 7 above)
Comment (zouyx) [reply filter/filter_impl/auth/provider_auth_test.go]: this comment has tag which is called outdated so...you must go to File Changed see the lasted update

File: filter/auth_spi.go [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/protocol"
+)
+
+type Authenticator interface {
+	Sign(protocol.Invocation, *common.URL) error
+	Authenticate(protocol.Invocation, *common.URL) error
+}
+
+type AccesskeyStorage interface {
```
Comment (zouyx) [inline filter/auth_spi.go]: why Accesskey here? But AccessKey in filter/filter_impl/auth/accesskey_storage.go

File: filter/auth_spi.go [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package filter
+
```
[the remark this answers is not in the document]
Comment (CodingSinger) [reply filter/auth_spi.go]: Maybe auth.go is better?

File: filter/auth_spi.go (hunk 10 above)
Comment (CodingSinger) [reply filter/auth_spi.go]: I will put `Accesskey` in auth_spi.go

File: filter/auth_ext.go [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+	Creator      string `yaml:"creator"   json:"creator,omitempty" property:"creator"`
+	Options      string `yaml:"options"   json:"options,omitempty" property:"options"`
+}
+
+type Authenticator interface {
+	Sign(protocol.Invocation, *common.URL) error
+	Authenticate(protocol.Invocation, *common.URL) error
+}
+
+type AccesskeyStorage interface {
```
Comment (zouyx) [inline filter/auth_ext.go]: i mean this class name:```AccesskeyStorage ``` key is lower case,but ```AccessKeyPair``` key is upper case

File: filter/filter_impl/auth/accesskey_storage.go [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/common/extension"
+	"github.com/apache/dubbo-go/filter"
+	"github.com/apache/dubbo-go/protocol"
+)
+
+type DefaultAccesskeyStorage struct {
```
Comment (zouyx) [inline filter/filter_impl/auth/accesskey_storage.go]: this key is lower case also.

File: filter/auth_spi.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (hxmhlt) [reply filter/auth_spi.go]: > Maybe auth.go is better?

Sounds good for me.

File: common/constant/key.go [hunk 14]
```diff
@@ -137,3 +137,21 @@ const (
...
+	ACCESS_KEY_STORAGE_KEY        = "accessKey.storage"
+	REQUEST_TIMESTAMP_KEY         = "timestamp"
+	REQUEST_SIGNATURE_KEY         = "signature"
+	AK_KEY                        = "ak"
+	SIGNATURE_STRING_FORMAT       = "%s#%s#%s#%s"
+	PARAMTER_SIGNATURE_ENABLE_KEY = "param.sign"
+	CONSUMER                      = "consumer"
+	ACCESS_KEY_ID_KEY             = "accessKeyId"
+	SECRET_ACCESS_KEY_KEY         = "secretAccessKey"
+)
```
Comment (hxmhlt) [inline common/constant/key.go]: I think auth is a optional selection for user. So it's not necessory for user to load these constant defination by 'extension'. I think these constant can move to filter_impl/auth folder

File: common/url.go [hunk 15]
```diff
@@ -313,6 +313,26 @@ func (c URL) ServiceKey() string {
...
+	}
+	group := c.GetParam(constant.GROUP_KEY, "")
+	buf.WriteString(":")
+	if group != "" {
+		buf.WriteString(group)
+	}
+	return buf.String()
+}
+
 func (c *URL) EncodedServiceKey() string {
```
Comment (hxmhlt) [inline common/url.go]: Why not use ServiceKey() func in url.go

File: filter/auth_ext.go [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+	GetAccesskeyPair(protocol.Invocation, *common.URL) *AccessKeyPair
+}
```
Comment (hxmhlt) [inline filter/auth_ext.go]: Move to filter_impl/auth/ folder?

Comment (flycash) [reply filter/auth_ext.go]: > Move to filter_impl/auth/ folder?

That is a good idea.

Comment (CodingSinger) [reply filter/auth_ext.go]: If move to auth folder, will  issue circular dependencies problem.

File: common/url.go (hunk 15 above)
Comment (CodingSinger) [reply common/url.go]: As far as I know, many places including the latest version of dubbo admin use the `ColonSeparatedKey` method to get the unique  key of the url.

File: common/constant/key.go (hunk 14 above)
Comment (CodingSinger) [reply common/constant/key.go]: ServiceConfig will need these constant variables.

File: filter/auth_ext.go (hunk 12 above)
Comment (CodingSinger) [reply filter/auth_ext.go]: ok, I will fix it.

File: filter/auth_ext.go (hunk 16 above)
Comment (CodingSinger) [reply filter/auth_ext.go]: I found that the relevant interfaces of TPS `TpsLimitStrategy` and` TpsLimitStrategyCreator` are placed in the `filter` directory.

File: common/constant/key.go (hunk 14 above)
Comment (hxmhlt) [reply common/constant/key.go]: I can not find these constant variables called by ServiceConfig in  your pr?

File: common/url.go (hunk 15 above)
Comment (hxmhlt) [reply common/url.go]: I means that ServiceKey method is designed for getting the unique key of the url.

Comment (hxmhlt) [reply common/url.go]: Is the method ServiceKey is not satisfy for you?

File: filter/auth_ext.go (hunk 16 above)
Comment (hxmhlt) [reply filter/auth_ext.go]: Why issue circular dependencies? I only found it called by filter/filter_impl/auth/authenticator.go.

Comment (hxmhlt) [reply filter/auth_ext.go]: Is there any other code not included in this pr?

File: common/constant/key.go (hunk 14 above)
Comment (CodingSinger) [reply common/constant/key.go]: Yes, I will add soon.

**Comment by zouyx** [comment]
Could you add some comment for public struct and method?

File: common/url.go [hunk 17]
```diff
@@ -339,6 +339,26 @@ func (c URL) ServiceKey() string {
 	return buf.String()
 }
 
+func (c *URL) ColonSeparatedKey() string {
```
Comment (zouyx) [inline common/url.go]: add some comment

**Pull request merged**